app/models/concepts_graph.rb

```diff
@@ -0,0 +1,98 @@
+# ConceptGraph transforms a track configuration to a form compatible for manipulating for
+# displaying the progression of the track
+class ConceptsGraph
+  attr_reader(
+    :language,
+    :nodes,
+    :edges,
+    :node_lookup_by_concept,
+    :node_lookup_by_index,
+    :adjacency,
+    :has_cycle
+  )
+
+  def initialize(track_config)
+    @language = track_config['language']
+    @nodes = make_nodes(track_config['exercises'])
+    @node_lookup_by_concept = make_concept_lookup
+    @node_lookup_by_index = make_index_lookup
```

Comment:
These can be deleted if extracted into the memoized method.

```suggestion
    @node_lookup_by_concept = make_concept_lookup
    @node_lookup_by_index = make_index_lookup
```